src/sentry/issues/issue_velocity.py

```diff
@@ -0,0 +1,131 @@
+                "past_week_event_count",
+            ),
+            Function(
+                "if",
+                [
+                    Function(
+                        "less", [Column("first_seen"), one_week_ago]
+                    ),  # if the issue is older than a week
+                    Function(
+                        "divide", [Column("past_week_event_count"), WEEK_IN_HOURS]
+                    ),  # divide the number of events in the week by a week in hours
+                    Function(
+                        "divide",
+                        [
+                            Column("past_week_event_count"),
+                            Function(
+                                "dateDiff", ["hour", Column("first_seen"), now]
+                            ),  # otherwise divide by its age in hours
+                        ],
+                    ),
+                ],
+                "hourly_event_rate",
+            ),
+        ],
+        groupby=[Column("group_id")],
+        where=[
+            Condition(
+                Column("timestamp"), Op.GTE, ninety_days_ago
+            ),  # include issues up to the oldest retention date to determine whether an issue is older than the week or not
+            Condition(Column("timestamp"), Op.LT, now),
+            Condition(Column("project_id"), Op.EQ, project.id),
+        ],
+        having=[
+            Condition(Column("past_week_event_count"), Op.GT, 1)
+        ],  # exclude any issues that had only 1 event in the past week
+        orderby=[OrderBy(Column("first_seen"), Direction.ASC)],
+        limit=Limit(10000),
+    )
+
+    query = Query(
+        match=subquery,
+        select=[
+            Function("quantile(0.9)", [Column("hourly_event_rate")], "p90")
+        ],  # get the approximate 90th percentile of the event frequency in the past week
+        limit=Limit(1),
+    )
+
+    request = Request(
+        dataset=Dataset.IssuePlatform.value,
+        app_id=REFERRER,
+        query=query,
+        tenant_ids={"referrer": REFERRER, "organization_id": project.organization.id},
+    )
+
+    result = raw_snql_query(request, referrer=REFERRER)["data"]
+    if len(result) == 0:
+        return None
+
+    try:
+        return result[0]["p90"]
+    except KeyError:
+        return None
+
+
+def set_velocity_threshold_for_project(project: Project) -> None:
+    threshold = calculate_velocity_threshold_for_project(project)
+    if threshold is None:
+        logger.error("Velocity threshold couldn't be calculated", extra={"project_id": project.id})
+        return
+    elif math.isnan(threshold):
+        pass  # TODO
+    # store in redis
+    with sentry_sdk.start_span(op="cluster.{CLUSTER_KEY}.set_velocity_threshold_for_project"):
+        client = redis_clusters.get(CLUSTER_KEY)
+        client.set(str(project.id), f"{threshold}")
+
+
+def get_velocity_threshold_for_project(project: Project):
```

Comment:
@isabellaenriquez @wedamija do we need a function that will check the cache and if there is a cache miss we fetch the value from snuba and store it? Or is that going to be handled separately?

Reply:
i think it'd make sense to check that here!

Reply:
I added what I think is the correct logic below for getting and setting in Redis based off other parts of the codebase where Redis is used, so (if you're reading this and are more familiar with Redis) let me know if I went about it wrong or if anything important is missing!